### PR TITLE
Deterministic query plans

### PR DESCRIPTION
**Pull request by ArturGajowy**
for a selection of queries.

The scope of this PR is providing fixes for the 3 queries included in tests and discussing strategy and providing groundwork for making the planner + optimizer duo deterministic (resulting in same plan for same query for a given presto release).

This addresses #6579

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -153,7 +155,7 @@
     private final boolean isDescribe;
 
     private final IdentityHashMap<FunctionCall, Signature> resolvedFunctions = new IdentityHashMap<>();
-    private final Set<SubqueryExpression> scalarSubqueries = newIdentityHashSet();
+    private final Set<Reference<SubqueryExpression>> scalarSubqueries = new LinkedHashSet<>();
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: A prominent source of non-determinisms are HashMaps and HashSets, which have a non-deterministic iteration order. This particular `identityHashSet` (replaced) was breaking TPC-DS Q6. I'd like to replace all of the `IdentityHashSet`-s and `IdentityHashMap`-s with wrapper Sets/Maps handling the wrapping and unwrapping of elements into Guava's [Equivalence.Wrapper<T>](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Equivalence.Wrapper.html). Otherwise, all usages of those identity-collections would have to be cluttered with manual wrapping and unwrapping. I'll provide an implementation of the collection soon for further discussion. For now, please let me know if the general idea sounds good :)

Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: @martint WDYT?

**Comment by ArturGajowy** [comment]
The commit order seems to be date-based on github... Will rewrite it so that parents are older than children :)

File: presto-tests/src/main/java/com/facebook/presto/tests/PlanDeterminismChecker.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        this(localQueryRunner, Function.identity());
+    }
+
+    public PlanDeterminismChecker(LocalQueryRunner localQueryRunner, Function<String, String> planEquivalenceFunction)
+    {
+        this.localQueryRunner = localQueryRunner;
+        this.planEquivalenceFunction = planEquivalenceFunction;
+    }
+
+    public void checkPlanIsDeterministic(String sql)
```
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/PlanDeterminismChecker.java:44]: s/check/assert ?

File: presto-tests/src/test/java/com/facebook/presto/tests/TestQueryPlansDeterministic.java [hunk 3]
```diff
@@ -82,4 +82,15 @@ public void testTpchQ9deterministic()
                 "  nation,\n" +
                 "  o_year DESC\n");
     }
+
+    @Test
```
Comment (kokosing) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestQueryPlansDeterministic.java:86]: I think you need to sort commits as this commit does not look as a second one.

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+ * for the underlying Map's keys.
+ * <p>
+ * EquivalenceMap instances retain iteration order of the underlying Map.
+ * Collections returned from #entrySet() and #keySet() are unmodifiable
+ * views.
+ *
+ * @param <K>
+ * @param <V>
+ */
+public class EquilvalenceMap<K, V>
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java]: can you please add some unit tests

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    private final Equivalence<Object> equivalenceRelation;
+
+    public static <K, V> EquilvalenceMap<K, V> on(MapSupplier supplier, Equivalence<Object> equilvalenceRelation)
+    {
+        return new EquilvalenceMap<>(supplier, equilvalenceRelation);
+    }
+
+    protected EquilvalenceMap(MapSupplier supplier, Equivalence<Object> equivalenceRelation)
+    {
+        this.delegate = supplier.supply();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java]: why supplier if you retrieve the value immediately?

**Comment by ArturGajowy** [comment]
The test for all `AbstractTestQueries` having a deterministic query plan is here: https://github.com/Teradata/presto/compare/deterministic_query_plans...Teradata:test_local_queries_deterministic

I didn't include it in this PR because the implementation is quite ugly and I couldn't find a better way. 
We can include it during the review, though :)

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+ * views.
+ *
+ * @param <K>
+ * @param <V>
+ */
+public class EquilvalenceMap<K, V>
+        implements Map<K, V>
+{
+    private final Map<Equivalence.Wrapper<K>, V> delegate;
+    private final Equivalence<Object> equivalenceRelation;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java]: why did you add suffix `Relation`?

**Review by kokosing — APPROVED** [review]
minor comments, please add for ATQ queries

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java (hunk 6 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java]: After a second reading I agree it's redundant

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java (hunk 5 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java]: So that the user does not have to know about the key type in the wrapped Map. Otherwise the constuctor would read:

`EquilvalenceMap(Map<Equivalence.Wrapper<K>, V> delegate, Equivalence<Object> equivalenceRelation)`.

The MapSupplier infers the type. And, thanks to the MapSupplier not knowing the key (and value) type in advance, there's less risk that users will pass a non-empty Map.

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java (hunk 4 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java]: Added some unit tests :)

File: presto-tests/src/main/java/com/facebook/presto/tests/PlanDeterminismChecker.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        IntStream.range(1, MINIMUM_SUBSEQUENT_SAME_PLANS)
+                .mapToObj(attempt -> getPlanText(sql))
+                .map(planEquivalenceFunction)
+                .reduce((s, s2) -> {
```
Comment (losipiuk) [inline presto-tests/src/main/java/com/facebook/presto/tests/PlanDeterminismChecker.java]: use more meaningful lambda arg names. E.g.
- expected
- candidate

File: presto-tests/src/test/java/com/facebook/presto/tests/TestQueryPlansDeterministic.java [hunk 8]
```diff
@@ -82,4 +82,15 @@ public void testTpchQ9deterministic()
...
+    public void testJoinsDeterministic()
+            throws Exception
+    {
+        determinismChecker.checkPlanIsDeterministic("" +
```
Comment (losipiuk) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestQueryPlansDeterministic.java]: maybe also test join without aggregation.

File: presto-tests/src/test/java/com/facebook/presto/tests/TestQueryPlansDeterministic.java [hunk 9]
```diff
@@ -93,4 +93,23 @@ public void testJoinsDeterministic()
                 "AND ps.brand = l.address"
         );
     }
+
+    @Test
+    public void testTpcdsQ6deterministic()
```
Comment (losipiuk) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestQueryPlansDeterministic.java:87]: I think it should `testTpcdsQ6Deterministic`

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+ * @param <K>
+ * @param <V>
+ */
+public class EquivalenceMap<K, V>
+        implements Map<K, V>
+{
+    private final Map<Equivalence.Wrapper<K>, V> delegate;
+    private final Equivalence<Object> equivalence;
+
+    public static <K, V> EquivalenceMap<K, V> on(MapSupplier supplier, Equivalence<Object> equivalence)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java]: Use `Supplier<Map<K,V>>`

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    {
+        return new EquivalenceMap<>(supplier, equivalence);
+    }
+
+    protected EquivalenceMap(MapSupplier supplier, Equivalence<Object> equivalence)
+    {
+        this.delegate = supplier.supply();
+        checkArgument(!(delegate instanceof SortedMap), "SortedMap delegates are not supported " +
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java]: This is not meaningful. Either leave just `"SortedMap delegates are not supported" or provide more verbose explanation (preferably as a comment, not exception message).

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    @Override
+    public V get(Object key)
+    {
+        return delegate.get(equivalence.wrap(key));
+    }
+
+    @Override
+    public V put(@Flow(target = "this.keys", targetIsContainer = true) K key, @Flow(target = "this.values", targetIsContainer = true) V value)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java]: I am pretty sure we do not want IntelliJ specific annotations in the codebase.
Or maybe it is fine. We use `org.intellij.lang.annotations.Language` a few times.

**Review by losipiuk — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java (hunk 10 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java]: This can't be a `Supplier<Map<K,V>>`, because the underlying Map is `Map<Equivalence.Wrapper<K>, V>`. Notice that MapSupplier's only method is generic on return type but the MapSupplier itself is not. This has the desired consequences descibed in my comment here: https://github.com/Teradata/presto/pull/413#discussion_r87391666

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java (hunk 5 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/util/maps/EquilvalenceMap.java]: Actually I guess it even is impossible to write a MapSupplier that will compile/not give warnings and will return a non-empty Map. Since the types aren't known in advance, the only possible return value is an empty Map (and this is thanks to type erasure).

File: presto-tests/src/main/java/com/facebook/presto/tests/PlanDeterminismChecker.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.stream.IntStream;
+
+import static org.testng.Assert.assertEquals;
+
+public class PlanDeterminismChecker
+{
+    private static final int MINIMUM_SUBSEQUENT_SAME_PLANS = 10;
+
+    private final LocalQueryRunner localQueryRunner;
+    private final Function<String, String> planEquivalenceFunction;
```
Comment (sopel39) [inline presto-tests/src/main/java/com/facebook/presto/tests/PlanDeterminismChecker.java:31]: Why is this needed? Can't you simply ensure that all query executions produce the same string plan? It would make the mechanism simpler.

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java (hunk 10 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java]: Do we need this supplier at all? Can't you just use `LinkedHashMap` as underlying structure?

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+ * Collections returned from #entrySet() and #keySet() are unmodifiable
+ * views.
+ * <p>
+ * Similarly to java.util.IdentityHashMap this class violates parts of
+ * Map's contract. See java.util.IdentityHashMap's javadoc for details.
+ *
+ * @param <K>
+ * @param <V>
+ */
+public class EquivalenceMap<K, V>
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java]: Why not call it `IdentityLinkedHashMap` (with identity equivalence) and use it whenever `IdentityHashMap` was used? Code structure would be simpler. Do we need supplier mechanism at all for now?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 15]
```diff
@@ -59,36 +60,36 @@
     private final List<Expression> parameters;
     private String updateType;
 
-    private final IdentityHashMap<Table, Query> namedQueries = new IdentityHashMap<>();
+    private final IdentityMap<Table, Query> namedQueries = new IdentityMap<>(LinkedHashMap::new);
 
-    private final IdentityHashMap<Node, Scope> scopes = new IdentityHashMap<>();
-    private final Set<Expression> columnReferences = newIdentityHashSet();
+    private final IdentityMap<Node, Scope> scopes = new IdentityMap<>(LinkedHashMap::new);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: I would be nicer (and simpler) to have `IdentityLinkedHashMap` class defined with empty constructor instead of composite `new IdentityMap<>(LinkedHashMap::new);`

File: presto-tests/src/test/java/com/facebook/presto/tests/TestLocalQueriesDeterministic.java [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+    protected MaterializedResult computeExpected(@Language("SQL") String sql, List<? extends Type> resultTypes)
+    {
+        determinismChecker.checkPlanIsDeterministic(sql);
+        return super.computeExpected(sql, resultTypes);
+    }
+
+    //Ignore queries that can't be used for plan determinism checking:
+
+    @Override
+    public void testDescribeInput()
```
Comment (sopel39) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestLocalQueriesDeterministic.java:208]: It would be great to classify tests (e.g: add group for tests that can be used to ensure deterministic planning). This kind of tests selection is very fragile (e.g: other will add irrelevant tests that will be executed here) and hard to maintain.

File: presto-tests/src/test/java/com/facebook/presto/tests/TestLocalQueriesDeterministic.java [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+        sessionPropertyManager.addConnectorSessionProperties(new ConnectorId(TESTING_CATALOG), TEST_CATALOG_PROPERTIES);
+
+        return localQueryRunner;
+    }
+
+    @Override
+    protected MaterializedResult computeActual(@Language("SQL") String sql)
+    {
+        determinismChecker.checkPlanIsDeterministic(sql);
+        return super.computeActual(sql);
```
Comment (sopel39) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestLocalQueriesDeterministic.java:73]: This executes quary and produces actual results, which shouldn't be concern for `TestLocalQueriesDeterministic`. Maybe tests that require `computeActual` method should be excluded to reduce overhead?

**Comment by sopel39** [comment]
I have pushed to sprint 44. However, I did not fix comments

File: presto-tests/src/main/java/com/facebook/presto/tests/PlanDeterminismChecker.java (hunk 7 above)
Comment (ArturGajowy) [reply presto-tests/src/main/java/com/facebook/presto/tests/PlanDeterminismChecker.java]: done

File: presto-tests/src/test/java/com/facebook/presto/tests/TestQueryPlansDeterministic.java (hunk 8 above)
Comment (ArturGajowy) [reply presto-tests/src/test/java/com/facebook/presto/tests/TestQueryPlansDeterministic.java]: removed the test as after rebasing on master it didn't catch any errors any more

File: presto-tests/src/main/java/com/facebook/presto/tests/PlanDeterminismChecker.java (hunk 13 above)
Comment (ArturGajowy) [reply presto-tests/src/main/java/com/facebook/presto/tests/PlanDeterminismChecker.java:31]: This is convenient for debugging. If I'm in the middle of investigation, it's super-easy to drop in a `plan -> plan.reaplaceAll("_\d+", "")` equivalence function, filtering out all the noise so that I can focus on query plan shape changes

File: presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java (hunk 14 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/util/maps/EquivalenceMap.java]: Done

File: presto-tests/src/test/java/com/facebook/presto/tests/TestLocalQueriesDeterministic.java (hunk 17 above)
Comment (ArturGajowy) [reply presto-tests/src/test/java/com/facebook/presto/tests/TestLocalQueriesDeterministic.java:73]: They're not a big overhead, all tests in `TestLocalQueriesDeterministic` run in under 60 s.

**Comment by ArturGajowy** [comment]
Rebased, applied comments. Removed the `EquivalenceMap` abstraction in favor of `IdentityLinkedHashMap` concretion as suggested by @sopel39.

**Comment by sopel39** [comment]
LGTM

**Comment by ArturGajowy** [comment]
Closing in favor of upstream PR
